politeiavoter/politeiavoter.go

```diff
@@ -615,6 +658,47 @@ func (c *ctx) _vote(token, voteId string) ([]string, *v1.BallotReply, error) {
 		return nil, nil, fmt.Errorf("no eligible tickets found")
 	}
 
+	// Put cast votes in a map so that we can
+	// filter in linear time
+	castVotes := make(map[string]v1.CastVote)
+	for _, v := range vrr.CastVotes {
+		castVotes[v.Ticket] = v
+	}
+
+	// Filter out tickets that have already voted
+	filtered := make([]*pb.CommittedTicketsResponse_TicketAddress, 0,
+		len(ctres.TicketAddresses))
+	for _, t := range ctres.TicketAddresses {
+		h, err := chainhash.NewHash(t.Ticket)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		v, ok := castVotes[h.String()]
+		if ok {
+			// Ticket has already voted. Validate vote.
+			sig, err := hex.DecodeString(v.Signature)
+			if err != nil {
+				return nil, nil, err
+			}
+
+			msg := v.Token + v.Ticket + v.VoteBit
+			validated, err := verifyMessage(t.Address, msg,
+				base64.StdEncoding.EncodeToString(sig))
+			if err != nil {
+				return nil, nil, err
+			}
+			if !validated {
+				return nil, nil, fmt.Errorf("could not validate vote %v",
```

Comment:
This looks like a super easy attack vector to prevent votes to me.  If the server just sends an invalid signature, it will cause the entire voter to bail and not cast other votes.

Reply:
I should note, I'm not just referring to this line, rather the entire section.

Reply:
Agreed.

Reply:
Good point.  It's been fixed.

Reply:
The PR has been updated to log any errors that are caused by bad vote data from the Politeia server instead of returning the errors.  Errors caused by bad data from dcrwallet will still cause politeiavoter to exit.